c-s mixed operation

muzarski opened the pull request:
# Motivation
Cassandra-stress supports mixed workloads. It allows to run multiple kinds of operations. Allowed operations are:
- read
- write
- counter_read
- counter_write

The parameters supported by `mixed` command are:
- all of the common parameters (e.g. used by `read`)
- all of the `counter_write` parameters
- `ratio()` parameter based on which tool samples an operation to execute. E.g. `ratio(read=2, write=1)` means that there will be approximately 2 read operations per 1 write operation.
- `clustering` distribution parameter which tells the tool how many times to run the operation that just got sampled.

# Changes
- Prepared `settings` module for `MixedOperation`
- Implemented `MixedOperation`

piodul commented on src/bin/cql-stress-cassandra-stress/settings/command/mixed.rs:
```diff
@@ -0,0 +1,94 @@
+            Command::Read => MixedSubcommand::Read,
+            Command::Write => MixedSubcommand::Write,
+            Command::CounterRead => MixedSubcommand::CounterRead,
+            Command::CounterWrite => MixedSubcommand::CounterWrite,
+            _ => anyhow::bail!("Invalid command for mixed workload: {}", cmd),
+        };
+        let weight = weight.parse::<f64>()?;
+        Ok((command, weight))
+    }
+
+    fn do_parse(s: &str) -> Result<Self> {
+        // Remove wrapping parenthesis.
+        let arg = {
+            let mut chars = s.chars();
+            anyhow::ensure!(
+                chars.next() == Some('(') && chars.next_back() == Some(')'),
+                "Invalid operation ratio specification: {}",
+                s
+            );
+            chars.as_str()
+        };
+
+        let mut command_set = HashSet::<MixedSubcommand>::new();
+        let weights = arg
+            .split(',')
+            .map(|s| -> Result<(MixedSubcommand, f64)> {
+                let (command, weight) = Self::parse_command_weight(s)?;
+                anyhow::ensure!(
+                    !command_set.contains(&command),
+                    "{} command has been specified more than once",
+                    command
+                );
+                command_set.insert(command);
+                Ok((command, weight))
+            })
+            .collect::<Result<Vec<_>, _>>()?;
+
+        Self::new(&weights)
+    }
+}
```
Side note: this falls under the "custom syntax" mentioned in https://github.com/scylladb/cql-stress/issues/52 (no need to do anything right now).

piodul commented on src/bin/cql-stress-cassandra-stress/settings/command/mixed.rs:
```diff
@@ -0,0 +1,94 @@
+
+pub type OperationRatio = EnumeratedDistribution<MixedSubcommand>;
+
+impl OperationRatio {
```
Please include a short comment which describes the syntax.

piodul commented on src/bin/cql-stress-cassandra-stress/java_generate/distribution/enumerated.rs:
```diff
@@ -0,0 +1,36 @@
+}
+
+impl<T: Copy + std::fmt::Display> EnumeratedDistribution<T> {
+    pub fn new(items: &[(T, f64)]) -> Result<Self> {
```
Nit: take `items` by value (i.e. `items: Vec<(T, f64)>`). If the caller prepared a `Vec<(T, f64)>` and no longer uses it, it will let them avoid a copy. In other cases, there should be no problem for the caller copy.

piodul commented on src/bin/cql-stress-cassandra-stress/java_generate/distribution/enumerated.rs:
```diff
@@ -0,0 +1,36 @@
+use rand_distr::{Distribution, WeightedIndex};
+
+#[derive(Clone)]
+pub struct EnumeratedDistribution<T: Copy + std::fmt::Display> {
```
As a principle, trait bounds should be avoided as possible in struct/enum definitions, unless they are required for the struct/enum definition to compile. https://stackoverflow.com/a/66369912

piodul commented on src/bin/cql-stress-cassandra-stress/operation/mixed.rs:
```diff
@@ -0,0 +1,181 @@
+        workload_factory: RowGeneratorFactory,
+        stats: Arc<ShardedStats>,
+    ) -> Result<Self> {
+        let mixed_params = settings.command_params.mixed.as_ref().unwrap();
+        let read_statement =
+            prepare_read_statement(regular_table_name, &session, &settings).await?;
+        let counter_read_statement =
+            prepare_read_statement(counter_table_name, &session, &settings).await?;
+        let write_statement = WriteOperationFactory::prepare_statement(&session, &settings).await?;
+        let counter_write_statement =
+            CounterWriteOperationFactory::prepare_statement(&session, &settings).await?;
+        let max_operations = settings.command_params.common.operation_count;
+        let operation_ratio = Arc::new(mixed_params.operation_ratio.clone());
```
I'm not a fan of the approach... You are exposing some details of the existing operations and use them to implement `MixedOperation`.

I think that the operations should be self-contained. `MixedOperation` should just encapsulate them and not care about their internals apart from the fact that they implement some trait. Would it be possible to implement it like that?

muzarski replied:
I thought about introducing some trait, but don't we end up with the same performance issue as in https://github.com/scylladb/cql-stress/pull/34 (`async-trait` boxed futures)?

I haven't compared both of the approaches when it comes to performance, but to me, it looks exactly the same - we end up with per-operation allocation from boxing the futures. I can still do the comparison, if necessary.

piodul replied:
I see. AFAIK `async` in trait methods is going to be stabilized in Rust 1.75, so maybe we can use `async_trait` until that happens.

muzarski replied:
> I see. AFAIK `async` in trait methods is going to be stabilized in Rust 1.75, so maybe we can use `async_trait` until that happens.

They indeed got stabilized, but unfortunately such traits are not object-safe, and so cannot be used in trait objects. I think the issue still holds. Do you still think we should use `async_trait`?

muzarski commented on src/bin/cql-stress-cassandra-stress/java_generate/distribution/enumerated.rs:
```diff
@@ -0,0 +1,36 @@
+        })
+    }
+
+    pub fn sample(&self) -> T {
+        self.items[self.dist.sample(&mut rand::thread_rng())].0
+    }
```
We do not depend on any determinism here, so we don't need to use a `java_random::Random`.

In addition, `rand_distr::WeightedIndex::sample` accepts `&mut R` where `R: rand::Rng`. We would need to implement `rand::RngCore` for `java_random::Random` (which is impossible, so we would need to wrap it with yet another type).

I just decided to keep it simple and use the rng provided by `rand` crate.

roydahan commented:
@muzarski let's try to move this one forward, once we will have this option supported we wlll be able to replace some of the currentl longevities that are using cassandra-stress with cql-stress.

muzarski commented:
v2:
- addressed review comments
- rebased on top of https://github.com/scylladb/cql-stress/pull/68. Can be useful to review mentioned PR to see how the changes interact with `mixed` command.

piodul commented on src/bin/cql-stress-cassandra-stress/operation/mixed.rs:
```diff
@@ -0,0 +1,169 @@
+        self.stats.get_shard_mut().account_operation(ctx, &result);
+
+        if result.is_ok() {
+            self.current_operation_count = self.current_operation_count.saturating_sub(1);
```
Is there a particular reason here for using saturating_sub? We check for `self.current_operation_count == 0` earlier in the function, so the current operation count here should not be `0`, right?

muzarski replied:
Yes, it's pointless to have saturating_sub here. It remains from the previous version where we would have:
```rust
self.current_operation_count = self.clustering_distribution.next_i64() as usize;
```
instead of
```rust
self.current_operation_count =
    (self.clustering_distribution.next_i64() as usize).max(1);
```

In the previous version there was a possibility that `self.current_operation_count` was 0 after taking the sample. Now we are guaranteed that it's at least 1 (thanks to `usize::max` method).

I'll get rid of the `saturating_sub`.

piodul commented on src/bin/cql-stress-cassandra-stress/operation/mixed.rs:
```diff
@@ -0,0 +1,169 @@
+            operation_ratio: Arc::clone(&self.operation_ratio),
+            clustering_distribution: mixed_params.clustering.create(),
+            current_operation: MixedSubcommand::Read,
+            current_operation_count: 0,
```
How about `current_operation_remaining`? It will be more descriptive IMO.

piodul reviewed (commented):
Left some nitpicks, after fixing them and rebasing it will LGTM.
